test/core/rtsmith_toml_test.cpp

```diff
@@ -0,0 +1,374 @@
+        }
+        actions = {
+            acl_drop();
+            @defaultonly NoAction();
+        }
+    }
+
+    apply {
+        if (hdr.eth_hdr.isValid()) {
+            drop_table.apply();
+        }
+    })");
+    auto autoContext = SetUp("bmv2", "v1model");
+    auto &rtSmithOptions = RtSmith::RtSmithOptions::get();
+    rtSmithOptions.target = "bmv2"_cs;
+    rtSmithOptions.arch = "v1model"_cs;
+
+    const char *configFilePath = "test/configuration.toml";
+    rtSmithOptions.setFuzzerConfigPath(configFilePath);
+    // Check if the fuzzer configuration path is set (and is set correctly).
+    ASSERT_TRUE(rtSmithOptions.fuzzerConfigPath().has_value());
+    ASSERT_EQ(rtSmithOptions.fuzzerConfigPath().value(), configFilePath);
+
+    auto rtSmithResultOpt = P4::P4Tools::RtSmith::RtSmith::generateConfig(source, rtSmithOptions);
+    // Check if the `RtSmithResult` object is generated successfully.
+    ASSERT_TRUE(rtSmithResultOpt.has_value());
+
+    auto compilerResult =
+        P4::P4Tools::RtSmith::RtSmith::generateCompilerResult(source, rtSmithOptions);
+    // Check if the `compilerResult` object is generated successfully.
+    ASSERT_TRUE(compilerResult.has_value());
+
+    const auto *programInfo = P4::P4Tools::RtSmith::RtSmithTarget::produceProgramInfo(
+        compilerResult.value(), rtSmithOptions);
+    // Check if the `programInfo` object is generated successfully.
+    ASSERT_TRUE(programInfo != nullptr);
+
+    // Parse the fuzzer configurations from the TOML file.
+    toml::parse_result tomlConfig;
+    try {
+        tomlConfig = toml::parse_file(configFilePath);
+    } catch (const toml::parse_error &e) {
+        error("P4RuntimeSmith: Failed to parse fuzzer configuration file: %1%", e.what());
+    }
+    int maxEntryGenCntConfig;
+    int maxAttemptsConfig;
+    int maxTablesConfig;
+    std::vector<std::string> tablesToSkipConfig;
+    uint64_t thresholdForDeletionConfig;
+    size_t maxUpdateCountConfig;
+    uint64_t maxUpdateTimeInMicrosecondsConfig;
+    uint64_t minUpdateTimeInMicrosecondsConfig;
+    auto maxEntryGenCntNode = tomlConfig["maxEntryGenCnt"];
+    // Check if the node exists.
+    if (!!maxEntryGenCntNode) {
+        // Check if the node represents an integer, meaning whether the configuration provided is
+        // valid.
+        if (maxEntryGenCntNode.type() == toml::node_type::integer) {
+            maxEntryGenCntConfig = maxEntryGenCntNode.value<int>().value();
+        } else {
+            error("P4RuntimeSmith: The maximum number of entries to generate must be an integer.");
+        }
+    } else {
+        // If the node does not exist, throw an error that indicates the configuration must be
+        // provided.
+        error("P4RuntimeSmith: The maximum number of entries to generate must be provided.");
+    }
+    auto maxAttemptsNode = tomlConfig["maxAttempts"];
+    if (!!maxAttemptsNode) {
+        if (maxAttemptsNode.type() == toml::node_type::integer) {
```

Comment:
This should probably be a helper function that uses templates. Since this just does casting and throws an error. Try to factor out all instances of this.

Reply:
E.g. `T castTomlNode<T>(TomlInput input)`. This function throws a gtest fail if the cast does not work.

Reply:
It seems to me that functions like `castTomlNode` that returns non-void values may not be able to call gtest failure functions that throw _fatal_ failures (e.g., [`FAIL`](http://google.github.io/googletest/reference/assertions.html#FAIL)). On top of that, I conjecture that it may be a good idea to leverage this template function in `config.h` and `config.cpp` (that includes logic of throwing `error`s instead of gtest failures). Thus, I walked around this by returning std::optional<T> and throwing either errors or gtest failures in the callers.